### ISSUE TITLE
Feature: Support Logical Ordering for Specific String Literal Types

### ISSUE THREAD

**Issue by akibarika** [azat-io/eslint-plugin-perfectionist#408]
### Describe the rule

In some cases, string literal types in TypeScript are better represented in a specific logical order rather than an alphabetical one. For example:

```tsx
type Breakpoint = 'xs' | 'sm' | 'md' | 'lg' | 'xl'; // Logical order
```

In this case, the progression from '`xs`' (extra small) to '`xl`' (extra large) follows a clear semantic or domain-specific logic. However, the current sorting rules in `eslint-plugin-perfectionist` would reorder these alphabetically:

```tsx
type Breakpoint = 'lg' | 'md' | 'sm' | 'xl' | 'xs'; // Alphabetical order
```

This disrupts the intended meaning and reduces the readability of such types, which are often used in design systems, configuration objects, or domain-specific types.

### Code example

1. Design Systems: Breakpoints in CSS frameworks ('`xs`', '`sm`', '`md`', '`lg`', '`xl`').

2. User States: User states in a workflow ('pending', 'active', 'completed', 'archived').

3. Error Severity Levels: Severity levels ('low', 'medium', 'high', 'critical').

4. Task Priorities: Task priority levels ('low', 'normal', 'high', 'urgent').

5. Font size: `thin`, `light`, `regular`, `bold`.

### Validations

- [X] Read the [docs](https://perfectionist.dev).
- [X] Check that there isn't [already an issue](https://github.com/azat-io/eslint-plugin-perfectionist/issues) that reports the same bug to avoid creating a duplicate.

**Comment by hugop95** [comment]
Hi @akibarika 

This is similar to https://github.com/azat-io/eslint-plugin-perfectionist/issues/148 and https://github.com/azat-io/eslint-plugin-perfectionist/issues/98.

I like this feature a lot, but I'm having trouble understanding the real use case:
Let's use your example

```ts
type Breakpoint = 'xs' | 'sm' | 'md' | 'lg' | 'xl'; // Logical order
```

[Question] Are there multiple cases where a user would need to define this specific logical order? If not, why not just disable the rule with `eslint-disable` for this type?

**Comment by akibarika** [comment]
Hi @hugop95 , 

Thank you for the quick response!

I agree with your points, and I also think it’s worth implementing this feature for objects.

Here’s a workaround I’ve been using:

```json
"rules": {
  "perfectionist/sort-objects": [
    "warn",
    {
      "groups": ["xs", "sm", "md", "lg", "xl", "unknown"],
      "customGroups": {
        "xs": "xs",
        "sm": "sm",
        "md": "md",
        "lg": "lg",
        "xl": "xl"
      }
    }
  ]
}
```
It’s not a perfect solution and has some limitations, but it does work for now.

**Comment by hugop95** [comment]
> I don't think there is a way to automatically detect the appropriate logic to use based on code context alone.

There is actually a way if we consider that most "logical" groups are a set of specific keys: we can use the fact that users can pass multiple configurations to a single rule.

For `sort-objects` in particular (the rule that benefits the most from this), this means that we could pass
```ts
[
  {
    groups: ['r', 'g', 'b'],
    customGroups: [
      {
        groupName: 'r',
        elementNamePattern: 'r',
      },
      {
        groupName: 'g',
        elementNamePattern: 'g',
      },
      {
        groupName: 'b',
        elementNamePattern: 'b',
      },
    ],
    configurationFilters: {
      allElementsMatchPattern: '^r|g|b$',
    },
  },
  {}, // Fallback configuration with default rule options
],
```

This means that

```ts
type Type = {
  r: string;
  g: string;
  b: string;
}
```

would be correctly sorted by `rgc` logic, while other types would use the fallback configuration.

@OlivierZal What do you think?

**Comment by OlivierZal** [comment]
Hi @hugop95,

I agree with you that the best place to configure it should be the `customGroups`. Here are some considerations though:

First I think that the `semantic sorting` topic is legitimate: another common use case which comes to my mind is `min` | `max`.

However, I also see possible consistency issues with sorting.

e.g. when there are only `min` and `max`, that's ok:
```ts
interface Range {
  min: number
  max: number
}
```

but what is the expected behavior when it's mixed with other keys, e.g.
```
interface HTMLCustomElement {
  id: string // let's say it's configured to be first
  a: string
  b: string
  ...
  m: string
  max: string
  mid: string
  min: string
  p: string
  ...
  z: string
  other: string // let's say it's configured to be last
}

Should the `min` | `max`...
* be left as is because they're mixed with other keys?
* be gathered and constitute a new block to be moved...
** first?
** right after the keys configured to be first?
** right before the keys configured to be last?
** last?
** to the place of the alphabetical-first element (`max`) – breaks the strict alphabetical sorting
** to the place of the semantical-first element (`min`) – breaks the strict alphabetical sorting

Breaking the alphabetical order could make `perfectionist` miss his target, that's why I agree that `customGroup` is flexible enough to handle such a case, but with some adaptation maybe – given that the concerned groups always have a finite number of elements:

```ts
[
  {
    groups: ['id, 'range', 'unknown', 'other'],
    customGroups: [
      {
        groupName: 'range',
        elementNamePattern: '^min|max$',
        internalGroupOrder: ['min', 'max'],
      },
    ],
  },
],
```

```ts
[
  {
    groups: ['id, 'rgb', 'unknown', 'other'],
    customGroups: [
      {
        groupName: 'rgb',
        elementNamePattern: '^r|g|b$',
        internalGroupOrder: ['r', 'g', 'b'],
      },
    ],
  },
],
```

What do you think @hugop95?

**Comment by hugop95** [comment]
@OlivierZal 

I like your proposal a lot, I think that the complexity justifies implementing it as another issue (considering that my proposal and yours are independent).

You are essentially adding another layer of `grouping` inside a group. Rather than adding a `internalGroupOrder` attribute that is limited to name pattern matching, we can simply re-use `groups`:

Your last example could become 

```ts
interface HTMLCustomElement {
  id: string
  minIndoorTemperature: string
  maxIndoorTemperature: string
  minOutdoorTemperature: string
  a: string
  z: string
  other: string
}
```

```ts
[
  {
    groups: ['id', 'range', 'unknown', 'other'],
    customGroups: [
      {
        groupName: 'indoor', // Only used within the `range` group
        elementNamePattern: 'IndoorTemperature$'',
      },
      {
        groupName: 'outdoor', // Only used within the `range` group
        elementNamePattern: 'OutdoorTemperature$'',
      },
      {
        groupName: 'id',
        elementNamePattern: '^id$',
      },
      {
        groupName: 'range',
        elementNamePattern: '^(min|max)',
        groups: ['indoor', 'outdoor'],
      },
      {
        groupName: 'other',
        elementNamePattern: '^other$',
      },
    ],
  },
],
```

**Comment by OlivierZal** [comment]
@hugop95, you got my point, I think we are on the same page 😊

**Comment by hugop95** [comment]
At first glance unfortunately, I don't think that's not something we can easily add, this will deserve its own issue 🙂

**Comment by akibarika** [comment]
Thank you @hugop95 & @OlivierZal !!

I've just updated to the latest version, `4.2.0`, but it seems to not be working.

```
Value {"groups":["r","g","b"],"customGroups":{"r":"r","g":"g","b":"b"},"useConfigurationIf":{"allNamesMatchPattern":"^r|g|b$"}} should NOT have additional properties.
```

It looks like the change hasn’t been included in the latest release yet.

**Comment by hugop95** [comment]
The feature hasn't been released yet indeed. It will be here 4.3.0 soon.

**Comment by OlivierZal** [comment]
And thank only @hugop95, _he_ did the whole work!

**Comment by h0adp0re** [comment]
Thanks for this feature, it looks like exactly what I need! I’m just curious about your release process. Do you have a timeline for when this might be published?

At my workplace, we typically publish features right after merging a PR. The thinking is — if time and energy have been spent on building a feature, it makes sense to make it available as soon as possible for everyone to use. It seems to help keep things moving smoothly, but I understand there are different approaches depending on the project.

**Comment by hugop95** [comment]
@h0adp0re The release process is handled by @azat-io, who can better explain the approach for this project.

In general:
- Critical hotfixes are released ASAP.
- Low-impact bugs and features tend to be grouped together in a release. There is another [big feature](https://github.com/azat-io/eslint-plugin-perfectionist/pull/404) about to be merged, so both features might end up in the same release.
- I think you can expect the release to be shipped by the end of the week.

**Comment by h0adp0re** [comment]
@hugop95 Wow, great job on the alphabet. And thanks for the outline.

I've noticed this feature grouping behavior in other open-source projects before and I have to say I don't quite get it 😄 — what if the feature being waited after takes several weeks more to refine? I am not here to argue or anything, just curious about the rationale behind the indeterminate waiting.

**Comment by azat-io** [comment]
@h0adp0re Hello! I will publish the release today

**Comment by azat-io** [comment]
Added in a3ee3ffeb88a69937d563cb8ac877507aec3b47a.
Released in [v4.3.0](https://github.com/azat-io/eslint-plugin-perfectionist/releases/tag/v4.3.0).

**Comment by h0adp0re** [comment]
Thank you very much!
